server/service/src/apis/api_on_central/patient_name_store_join.rs

```diff
@@ -0,0 +1,157 @@
+
+    info!(
+        "Created name_store_join for patient {} and store {}",
+        name_id, store_id
+    );
+
+    Ok(())
+}
+
+async fn add_patient_to_central(
+    service_provider: &ServiceProvider,
+    ctx: &ServiceContext,
+    name_id: &str,
+) -> Result<(), CentralApiError> {
+    let central_store_id = ActiveStoresOnSite::get(&ctx.connection)?
+        .store_ids()
+        .first()
+        .ok_or_else(|| {
+            error!("No active stores found for this site");
+            CentralApiError::InternalError("Central server misconfigured".to_string())
+        })?
+        .to_owned();
+
+    // Add visibility for this patient to central site
+    link_patient_to_store(service_provider, ctx, &central_store_id, name_id).await?;
+
+    info!(
+        "Created name_store_join for patient {} and central store {}",
+        name_id, central_store_id
+    );
+
+    // TODO: possibly should check is not pre-initialisation here?
+    service_provider.sync_trigger.trigger(None);
+    info!("Sync cycle triggered to receive patient records");
+
+    wait_for_sync(service_provider, &ctx, name_id).await?;
+
+    Ok(())
+}
+
+async fn wait_for_sync(
+    service_provider: &ServiceProvider,
+    ctx: &ServiceContext,
+    name_id: &str,
+) -> Result<(), CentralApiError> {
+    let timeout = Duration::from_secs(30);
+    let start_time = tokio::time::Instant::now();
+
+    loop {
+        // Check if we've exceeded the timeout
+        if start_time.elapsed() > timeout {
+            error!("Timeout waiting for sync to complete");
+
+            return Err(CentralApiError::InternalError(
+                "Error adding patient visibility".to_string(),
+            ));
+        }
+
+        // Brief pause to avoid busy loop, and hopefully give time for sync to start
+        let duration = Duration::from_millis(500);
+        sleep(duration).await;
+
+        let sync_status = match service_provider
+            .sync_status_service
+            .get_latest_sync_status(ctx)?
+        {
+            Some(sync_status) => sync_status,
+            None => {
+                error!("Could not find latest sync log");
+                return Err(CentralApiError::InternalError(
+                    "Error adding patient visibility".to_string(),
+                ));
+            }
+        };
+
+        // Potential race condition, sync is triggered in separate process so may not
+        // have started syncing yet in first loop
+        // More robust to check patient record has been received
+        if !sync_status.is_syncing {
+            // If sync finished but integration of patient failed, will break after timeout
+            if check_patient_exists(&ctx.connection, &name_id)?.is_some() {
+                info!("Patient data received");
+                break;
+            }
+        }
+        debug!("Patient not yet found, awaiting sync completion...");
+    }
+
+    Ok(())
+}
+
+impl From<GetActiveStoresOnSiteError> for CentralApiError {
```

Comment:
At some point we agreed that `?` make it hard to trace the conversion, so should only convert error for common things. Not requesting change now, but I would do a helper method in CentralApiError (that accept same thing as format_error)

Reply:
Thanks, ended up replacing with map_error in the next PR, adding the processor 🙏